components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java

```diff
@@ -706,47 +732,231 @@ public static String getUserConsentURL(OAuth2Parameters params, String loggedInU
         return consentPage;
     }
 
-    private static String getFilteredScopes(OAuth2Parameters params) throws OAuthSystemException {
+    /**
+     * Deny consent for given set of OAuth scopes for the application.
+     *
+     * @param params    OAuth2 request parameters.
+     * @param user      Authenticated user.
+     * @throws OAuthSystemException
+     */
+    public static void handleDeniedOAuthScopeConsent(OAuth2Parameters params, AuthenticatedUser user)
+            throws OAuthSystemException {
 
         try {
-            Set<String> allowedScopes = params.getScopes();
-            startTenantFlow(params.getTenantDomain());
+            Set<String> userDeniedScopesSet = params.getConsentRequiredScopes();
+            if (CollectionUtils.isNotEmpty(userDeniedScopesSet)) {
+                List<String> disapprovedScopeConsents = new ArrayList<>(userDeniedScopesSet);
+                // Remove OIDC scopes.
+                disapprovedScopeConsents.removeAll(
+                        Arrays.asList(ArrayUtils.nullToEmpty(oAuthAdminService.getScopeNames())));
+                // Update user consent by merging existing consent.
+                String userId = getUserIdOfAuthenticatedUser(user);
+                String appId = getAppIdFromClientId(params.getClientId());
+                oAuth2ScopeService.updateUserConsentForApplication(userId, appId,
+                        IdentityTenantUtil.getTenantId(user.getTenantDomain()),
+                        null, disapprovedScopeConsents);
+            }
+        } catch (IdentityOAuthAdminException e) {
+            throw new OAuthSystemException(
+                    "Error occurred while removing OIDC scopes from disapproved OAuth scopes.", e);
+        } catch (IdentityOAuth2ScopeException e) {
+            throw new OAuthSystemException(("Error occurred while updating OAuth scope consents."));
+        }
+    }
+
+    /**
+     * Check if the user has already given consent to required OAuth scopes.
+     *
+     * @param user              Authenticated user.
+     * @param oAuth2Parameters  OAuth2 parameters.
+     * @return  True if user has given consent to all the requested OAuth scopes.
+     * @throws IdentityOAuth2ScopeConsentException
+     * @throws IdentityOAuthAdminException
+     */
+    public static boolean isUserAlreadyConsentedForOAuthScopes(AuthenticatedUser user,
+                                                               OAuth2Parameters oAuth2Parameters)
+            throws IdentityOAuth2ScopeException, IdentityOAuthAdminException, OAuthSystemException {
+
+        List<String> scopesToBeConsented = new ArrayList<>(oAuth2Parameters.getScopes());
+        // Remove OIDC scopes.
+        scopesToBeConsented.removeAll(Arrays.asList(ArrayUtils.nullToEmpty(oAuthAdminService.getScopeNames())));
```

Comment:
Since this is used in multiple places would be better to move to a private method like `getOIDCScopeNames()` 
`Arrays.asList(ArrayUtils.nullToEmpty(oAuthAdminService.getScopeNames()))`

Reply:
Fixed in https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1548/commits/27ddbc8fed9c4368ffe7a826b857537a6b8ffff0